### DISCUSSION TITLE
Partial Determinism

### DISCUSSION THREAD

**Discussion by rowanfr** [avianphysics/avian#197] [Ideas]
Would it be possible to implement determinism for only some physics objects, by that I mean have some physics objects which uniquely are deterministic so that those can be used for multi-platform games and enable easier netcode (by allowing for rollback) while having most other physics objects by non-deterministic. The deterministic objects would have to only be allowed to impart force to non-deterministic objects and not receive any force (otherwise they wouldn't be deterministic). This way a few gameplay relevant objects could retain determinism while allowing for other aesthetic objects to respond appropriately to these deterministic objects without substantial inefficiencies in speed for those aesthetic objects. I think this method of operating would fit an ECS style of development better then the Rapier either being deterministic or non-deterministic. Is this a practical direction or am I fundamentally misunderstanding determinism? Separately would this even be as efficient as global determinism?

**Comment by Jondolf** [comment]
Determinism in bevy_xpbd requires us to use certain features for our dependencies; `enhanced-determinism` for Parry's collision detection, and `libm` for Glam's math types. You can't have both the deterministic and non-deterministic versions of these in the same app at the same time because of conflicts.

Excluding these dependencies, bevy_xpbd should actually be just as deterministic with `enhanced-determinism` disabled as it is with the feature enabled. We don't currently need to do anything differently for determinism to work. This might change in the future if e.g. increasing parallelization needs extra work to support determinism, but for now, `enhanced-determinism` shouldn't have any extra overhead in terms of bevy_xpbd specifically.

But yeah, to have proper determinism for *any* entities we need to have the Parry and Glam features enabled, and this means that they are enabled everywhere. If we ignore this, we could probably treat entities set as deterministic/indeterministic differently, but I would suspect the difference to be almost negligible in practise, and it wouldn't currently be useful because we don't handle determinism any differently.

**Reply by rowanfr** [reply]
Thank you for letting me know, I'm fairly new to physics systems of all kinds so I'm mostly speaking as to how earlier one's have been with regards to performance.
